### PR TITLE
Inference support for `mps` device

### LINKED ISSUE #292: Add inference support for `mps` device (Apple Silicon)
There is a lot of community interest in running diffusers on Apple Silicon. A first step could be to introduce support for the `mps` device, which currently requires PyTorch nightly builds. Another step down the line could be to convert to Core ML and optimize to make sure that the use of the ANE (Neural Engine) is maximized.

This PR deals with the first approach.

**Describe the solution you'd like**
The following should work:

```Python
pipe = StableDiffusionPipeline.from_pretrained(model_id, use_auth_token=True)
pipe = pipe.to("mps")

# Rest of inference code remains the same
```

`pipe.to` would determine whether `mps` is available in the computer, and would raise an error if it's not.

One consideration is that perhaps not _all_ the components have to be moved to the `mps` device. Perhaps it's more efficient or practical to keep the text encoder in CPU, for instance. If so, `to()` would move the required components and the pipelines would need to be adapted to move the tensors transparently.

**Describe alternatives you've considered**
Conversion to Core ML, as mentioned above.

**Additional context**
I have tested the `unet` module in `mps` vs `cpu`, and these are some preliminary results in my computer (M1 Max with 64 GB of unified RAM), when iterating through the 51 default steps of Stable Diffusion with the default scheduler and no classifier-free guidance:

```
Device: cpu, torch: 1.12.1,  time: 92.09s
Device: cpu, torch: 1.13.0.dev20220830,  time: 92.12s
Device: mps:0, torch: 1.13.0.dev20220830,  time: 17.20s
```

Execution in `mps` complains about one operation having to be performed in CPU:

```
The operator 'aten::masked_select' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
```

We need to investigate the reason and whether an alternative would yield better performance.

### PR DESCRIPTION
**Pull request by pcuenca**
This addresses the issues identified when assessing inference on Apple Silicon, see https://github.com/huggingface/diffusers/issues/292#issuecomment-1236112207 for details.

### Current status

- Stable Diffusion pipeline works.
- Results in CPU / MPU are reproducible when using the same seeds. Generators do not work in the `mps` device, so we needed some minor adjustments.
- Some incompatible ops identified by failing tests were rewritten (or we fall back to CPU). But I have only verified `test_models_unet` and `test_models_vae`.
  - See f20a0dda6f1d2b46ac872852be4ad58188e63a5e
  - See 7f40f24297e4d3dc0279fdfe38ae4015de8b2747 and https://github.com/pytorch/pytorch/issues/84535
- Determinism tests pass, but the solution is a workaround hack.

### The hack

We perform a one-time "warmup" forward pass through `unet` and `vae` because the result from the first pass is different than subsequent results. I suspect something related to randomness might be at play, but I haven't identified the root cause. We have several options here:
- Remove the hack, it's probably overkill. We can recommend users to run a full pass through the pipeline (1 step is enough) after moving it to the device. Downside is that determinism tests fail in `test_models_unet` and `test_models_vae`.
- Find the cause and apply a proper solution.

I'd like to merge soon but don't like the hack. Can we remove it and live with some failing tests until we can find what's causing the issue?

Update: we only perform the hack during testing. Users are recommended to perform an initial pass if they care about reproducibility.

### To do:
- [x] Some `unet` and `vae` tests still fail. For example, `test_output_pretrained` in `AutoencoderKLTests`.
- [x] Translate some incompatible ops or fall back to CPU.
- [x] Ensure tests pass.
- [x] ~~Find a more principled workaround to perform the initial "warm up" pass~~. I tried to use forward hooks, but couldn't do it because they don't pass keyword arguments, which we use in many of our `forward` implementations.
- Create fail case for first-pass issue (results are different than in subsequent passes). Deferred to be tracked in #372.

Fixes #292.

### CODE REVIEW

File: src/diffusers/models/resnet.py
Comment (patrickvonplaten) [inline src/diffusers/models/resnet.py:450]: this deserves a comment here IMO -> are some kernels not yet supported on `"mps"`?

File: src/diffusers/models/vae.py
Comment (patrickvonplaten) [inline src/diffusers/models/vae.py:398]: Not a big fan of an additional abstraction - I'd like to avoid it here first because of the following:

- Makes the code harder to understand for everybody that's not using MPS
- Makes MPS somewhat hard to debug -> now when people code just one forward pass, actually two forward passes are done
- Too much black magic - I'd prefer to instead make sure our docs are good here instead

**Review by patrickvonplaten — COMMENTED** [review]
Not a big fan of the warmup abstraction - I'd prefer to just leave it to the user to call the warmup method. Ok with adding a warmup_mps() method to the `ModelMixin`, but don't like that this is automatically called - wdyt @pcuenca ?

**Comment by patrickvonplaten** [comment]
So I'd advocate for the following:

- Add an experimental `_mps_warmup(...)` method to `ModelMixin` and `PipelineMixin` that the user has to call themselves.
- With the underscore `_` and with a note we make sure to tell people it's experimental and due to a hack currently. 
- We can adapt the tests with some `if device is ` statements to make them pass

File: src/diffusers/models/resnet.py
Comment (pcuenca) [reply src/diffusers/models/resnet.py:450]: Yes, that causes a runtime error: https://github.com/pytorch/pytorch/issues/84535

**Comment by pcuenca** [comment]
> So I'd advocate for the following:
> 
> * Add an experimental `_mps_warmup(...)` method to `ModelMixin` and `PipelineMixin` that the user has to call themselves.
> * With the underscore `_` and with a note we make sure to tell people it's experimental and due to a hack currently.
> * We can adapt the tests with some `if device is ` statements to make them pass

Actually I think I'd recommend users to run a pipeline pass with 1 iteration instead before using the outputs, since that's easier. We can use the underscored methods ourselves just to make tests pass.

**Comment by pcuenca** [comment]
> Not a big fan of the warmup abstraction - I'd prefer to just leave it to the user to call the warmup method. Ok with adding a warmup_mps() method to the `ModelMixin`, but don't like that this is automatically called - wdyt @pcuenca ?

Yes, you are right, it's not important enough to expose it to all users. Thanks!

File: README.md
Comment (patil-suraj) [inline README.md:45]: Think we should move this under docs or examples. Maybe it makes sense now to add a doc section now that explains how to run diffusers with mps, onnx, and other hardwares. cc @anton-l @patrickvonplaten

File: README_mps.md
Comment (patil-suraj) [inline README_mps.md:36]: (nit)
```suggestion
image = pipe(prompt).images[0]
```

File: src/diffusers/modeling_utils.py
Comment (patil-suraj) [inline src/diffusers/modeling_utils.py:593]:
```suggestion
        rest, for the same inputs. We are investigating the root cause of the problem, but meanwhile these methods can be
```

File: src/diffusers/models/attention.py
Comment (patil-suraj) [inline src/diffusers/models/attention.py:140]: can't we always do `x.contiguous()` on other devices as well ?

File: src/diffusers/models/unet_2d_condition.py
Comment (patil-suraj) [inline src/diffusers/models/unet_2d_condition.py:139]: shouldn't this be handled in the `if` block also ?

also can we always cast it to `float32` cc @patrickvonplaten ?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:107]: this should not be removed here, this was added to get 1:1 results with stable diffusion cf https://github.com/huggingface/diffusers/pull/182#discussion_r945939774

File: src/diffusers/schedulers/scheduling_pndm.py
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_pndm.py:274]: do we need this in other schedulers as well for ex DDIM, K-LMS ?

File: tests/test_models_unet.py
Comment (patil-suraj) [inline tests/test_models_unet.py:196]: think we can always cast it to `int32`

File: README.md
Comment (pcuenca) [reply README.md:45]: Yeah I wasn't sure where to put it. `docs` sounds good to me, should we create a new section there?

File: README_mps.md
Comment (pcuenca) [reply README_mps.md:36]: Sure, this became outdated with respect to `main`, I'll merge and adapt. Thanks!

File: src/diffusers/models/attention.py
Comment (pcuenca) [reply src/diffusers/models/attention.py:140]: We can, but my approach was to be as transparent as possible with the way it works in other devices. In this case I also wanted to prevent a memory copy. In addition, we might want to explore "channels last" format in the future (not sure if it would help here), and I wasn't sure about the implications of using contiguous if we do.

But happy to change it if it's clearer that way.

File: src/diffusers/models/unet_2d_condition.py
Comment (pcuenca) [reply src/diffusers/models/unet_2d_condition.py:139]: I think I debugged the other branch and it worked fine (at least when using an `int`), I'll double check to make sure.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:107]: Ah, missed that, sorry. I saw that the behaviour in the rest of pipelines was different (`latents` are always moved to the device after generation, except here), so we might want to use a comment in this case.

Doing it like that doesn't work in `mps` (the `generator` is essentially ignored, I think), so for reproducibility, and to obtain the same results as when the CPU is used, we'll need to use a conditional in this case.

File: src/diffusers/schedulers/scheduling_pndm.py
Comment (pcuenca) [reply src/diffusers/schedulers/scheduling_pndm.py:274]: Not sure. Scheduler tests passed, but they might not be comprehensive enough. I tested LMS by configuring the pipeline manually, will run the others too.

File: README.md
Comment (patil-suraj) [reply README.md:45]: yeah a new section called `optimization` where we could put docs related to mps, onnx, openvino and AMD support etc.

File: tests/test_models_unet.py
Comment (pcuenca) [reply tests/test_models_unet.py:196]: The reason was to not interfere with the common path. I observed some slight differences when using `int32` instead of `int64` during the Flax tests, but they might not be relevant here.

File: src/diffusers/models/attention.py
Comment (patrickvonplaten) [reply src/diffusers/models/attention.py:140]: `x.contiguous()` allocates new memory, so happy trying to avoid it if possible :-)

File: src/diffusers/models/unet_2d_condition.py
Comment (patrickvonplaten) [reply src/diffusers/models/unet_2d_condition.py:153]: Yes I think we can always cast here indeed

File: tests/test_models_unet.py
Comment (patrickvonplaten) [reply tests/test_models_unet.py:196]: Both work fine with me (Flax will be in another file and surely not run on "mps")

File: README_mps.md
Comment (patrickvonplaten) [inline README_mps.md:13]: Maybe add torch version as a requirement to the section above. Nightly will change soon

File: README_mps.md
Comment (patrickvonplaten) [inline README_mps.md:33]: Do we need the "warmup" functions at all? They don't seem to be used here. Could we maybe just delete them in this case? For the tests we could do something like:

```python
if torch_device == "mps":
    pipe(prompt, num_inference_steps=1)
```

as well no? 

Think this would actually save quite some code and be cleaner no?

**Review by patrickvonplaten — COMMENTED** [review]
Thanks a lot @pcuenca !

After some thought and reading through your PR, I actually think we can remove the warmup functions completely no? 

Looking at your README_mps it seems like they aren't used at all and in this case let's just not add them (it's easier to understand the way it's done in the README anyways IMO)

Would it be ok to actually remove them all together?

File: README_mps.md
Comment (patrickvonplaten) [reply README_mps.md:33]: Very nice README !

**Comment by patrickvonplaten** [comment]
Also could we run all slow tests once before merging this as it touches many files? :-)

File: tests/test_models_unet.py
Comment (pcuenca) [reply tests/test_models_unet.py:196]: I meant it as a reference point, not to try to integrate it here :)

**Comment by pcuenca** [comment]
Thanks for the useful comments!

> After some thought and reading through your PR, I actually think we can remove the warmup functions completely no?

I'd love to. They are not exposed to users, but they are used in the following test functions:

- `test_from_pretrained_save_pretrained`: https://github.com/huggingface/diffusers/pull/355/files#diff-a6ebb0bccfe159ca855e9d9b5d6fbbeb62fced0c0296de0993ba39ef5c862cf0R34-R35 and lines below.
- `test_determinism`: https://github.com/huggingface/diffusers/pull/355/files#diff-a6ebb0bccfe159ca855e9d9b5d6fbbeb62fced0c0296de0993ba39ef5c862cf0R63-R64

The idea is that `pytest -x tests/test_models_vae.py` and `pytest -x tests/test_models_unet.py` would work, but they test models independently of an associated pipeline.

(Those tests currently don't pass after merging the new `BaseOutput` code from main, I'll fix them).

File: tests/test_modeling_common.py
Comment (pcuenca) [inline tests/test_modeling_common.py:222]: I added `with torch.no_grad():` because computing gradients does not work well in `mps`. Is it important to do it in  the other devices and we can disable them for all these tests? /cc @patrickvonplaten @patil-suraj

File: src/diffusers/schedulers/scheduling_pndm.py
Comment (pcuenca) [reply src/diffusers/schedulers/scheduling_pndm.py:284]: I verified them and updated `ScoreSdeVeScheduler`, the others do not use the same type of indexing and it works.

**Comment by pcuenca** [comment]
Fast tests now pass (except `test_full_loop_no_noise`, which is failing for me in `main` too).

**Comment by patrickvonplaten** [comment]
@pcuenca could you run `black` here as well? :-)

**Comment by pcuenca** [comment]
> @pcuenca could you run `black` here as well? :-)

Of course! I've been neglecting 'make style' because it changes all the files, so I run 'black' manually :)

**Comment by pcuenca** [comment]
Update: ran slow tests in a `cuda` device and they behave the same as in the current `main` (after #364 was merged). These are my results in both branches:

```
FAILED tests/test_models_unet.py::UNetLDMModelTests::test_output_pretrained - AssertionError: False is not true
FAILED tests/test_models_unet.py::NCSNppModelTests::test_output_pretrained_ve_mid - AssertionError: False is not true
FAILED tests/test_models_vae.py::AutoencoderKLTests::test_output_pretrained - AssertionError: False is not true
FAILED tests/test_models_vq.py::VQModelTests::test_output_pretrained - AssertionError: False is not true
FAILED tests/test_pipelines.py::PipelineTesterMixin::test_score_sde_ve_pipeline - AssertionError: assert 0.6438006442010402 < 0.01
FAILED tests/test_pipelines.py::PipelineTesterMixin::test_stable_diffusion_fast_ddim - AssertionError: assert 0.0050546875000000435 < 0.001
FAILED tests/test_pipelines.py::PipelineTesterMixin::test_stable_diffusion_img2img_pipeline - AssertionError: assert 0.7294117647058824 < 0.0001
FAILED tests/test_pipelines.py::PipelineTesterMixin::test_stable_diffusion_img2img_pipeline_k_lms - AssertionError: assert 0.5882352941176471 < 0.0001
FAILED tests/test_pipelines.py::PipelineTesterMixin::test_stable_diffusion_inpaint_pipeline - AssertionError: assert 0.03921568627450989 < 0.001
```

**Comment by pcuenca** [comment]
I think this is good for a final review. I added "Known Issues" to `README_mps` (which I propose to move to the docs during our sprint tomorrow) with these items:

> - As mentioned above, we are investigating a strange [first-time inference issue](https://github.com/huggingface/diffusers/issues/372).
> - Generating multiple prompts in a batch [crashes or doesn't work reliably](https://github.com/huggingface/diffusers/issues/363). We believe this might be related to the [`mps` backend in PyTorch](https://github.com/pytorch/pytorch/issues/84039#issuecomment-1237735249), but we need to investigate in more depth. For now, we recommend to iterate instead of batching.

Is there anything else we are missing?

File: src/diffusers/modeling_utils.py
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:603]: @pcuenca sorry I think my previous comments weren't super clear here. 
Do you think we could maybe remove this method all together as it's now only used in the tests and I think adding functionality to the core `modeling_utils.py` function is not really worth it just to make the tests easier. Could we maybe instead just add a generic `warm_up(...)` functions just in the testing files so that it doesn't go into the `pip` package at all?

Think it would be very nice to add as little code as possible to the general `Mixin` class in the beginning to not scare readers. 

Does this make sense?

File: src/diffusers/models/resnet.py
Comment (patrickvonplaten) [inline src/diffusers/models/resnet.py:451]: think this doesn't really warrant a new variable -> it now forces me to look up what `using_mps` means every time I see it

File: src/diffusers/models/unet_2d.py
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d.py:205]: this is is only used in testing - could we remove the function maybe?

File: src/diffusers/models/vae.py
Comment (patrickvonplaten) [inline src/diffusers/models/vae.py:452]: Also here can we remove the function and only add it to the test files?

File: src/diffusers/models/vae.py
Comment (patrickvonplaten) [inline src/diffusers/models/vae.py:533]: Can we also remove it and only use in test files?

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:143]: Let's maybe use this unit test decorator instead:

```
@unittest.skipIf(torch_device == "mps", "Training is not supported on Mac yet")
```

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:145]: Let's maybe try to use the unit test decorator for consistency with other tests

File: src/diffusers/models/resnet.py
Comment (patrickvonplaten) [inline src/diffusers/models/resnet.py:458]:
```suggestion
    if out.device.type == "mps":
```

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:23]:
```suggestion
```

Let's just use unittest to skip the tests :-)

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:36]: Can we write a helper function right into the tests here to not have to add testing functionality to core `src/diffusers` (which would go into pip)

**Comment by patrickvonplaten** [comment]
Also @pcuenca - could you downgrade black to 22.3.0 so that `make style` works? :-) @anton-l let's upgrade black after the release on Thursday :-)

File: src/diffusers/models/resnet.py
Comment (pcuenca) [reply src/diffusers/models/resnet.py:458]: No longer true, it was moved to `cpu` for the fallback computation.

File: src/diffusers/modeling_utils.py
Comment (pcuenca) [reply src/diffusers/modeling_utils.py:603]: Yes, I agree, makes sense. I inadvertently got attached to the classes knowing the size of their inputs, it's unfortunate that we can't define those extensions separately in another file. I made the changes to the testing code but it took me a bit longer than expected because tests were giving inconsistent results for some reason.

File: src/diffusers/models/resnet.py
Comment (pcuenca) [reply src/diffusers/models/resnet.py:458]: Using `input` as the reference for the device instead.

File: tests/test_modeling_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_common.py:45]: nice!

**Review by patrickvonplaten — APPROVED** [review]
Looks very nice! Feel free to merge whenever @pcuenca :-)

**Comment by pcuenca** [comment]
Thanks for the help!

**Pull request merged**
